File: src/main/java/net/atos/zac/app/zoeken/converter/RESTZaakZoekObjectConverter.java
/*
 * SPDX-FileCopyrightText: 2022 Atos
 * SPDX-License-Identifier: EUPL-1.2+
 */

package net.atos.zac.app.zoeken.converter;

import java.util.HashMap;

import jakarta.inject.Inject;

import net.atos.zac.app.policy.converter.RestRechtenConverter;
import net.atos.zac.app.zoeken.model.RESTZaakZoekObject;
import net.atos.zac.policy.PolicyService;
import net.atos.zac.util.time.DateTimeConverterUtil;
import net.atos.zac.zoeken.model.ZaakIndicatie;
import net.atos.zac.zoeken.model.zoekobject.ZaakZoekObject;

public class RESTZaakZoekObjectConverter {

    @Inject
    private PolicyService policyService;

    public RESTZaakZoekObject convert(final ZaakZoekObject zoekItem) {
        final RESTZaakZoekObject restZoekItem = new RESTZaakZoekObject();
        restZoekItem.id = zoekItem.getObjectId();
        restZoekItem.type = zoekItem.getType();
        restZoekItem.identificatie = zoekItem.getIdentificatie();
        restZoekItem.omschrijving = zoekItem.getOmschrijving();
        restZoekItem.toelichting = zoekItem.getToelichting();
        restZoekItem.archiefNominatie = zoekItem.getArchiefNominatie();
        restZoekItem.archiefActiedatum = DateTimeConverterUtil.convertToLocalDate(zoekItem.getArchiefActiedatum());
        restZoekItem.registratiedatum = DateTimeConverterUtil.convertToLocalDate(zoekItem.getRegistratiedatum());
        restZoekItem.startdatum = DateTimeConverterUtil.convertToLocalDate(zoekItem.getStartdatum());
        restZoekItem.einddatum = DateTimeConverterUtil.convertToLocalDate(zoekItem.getEinddatum());
        restZoekItem.einddatumGepland = DateTimeConverterUtil.convertToLocalDate(zoekItem.getEinddatumGepland());
        restZoekItem.uiterlijkeEinddatumAfdoening = DateTimeConverterUtil.convertToLocalDate(zoekItem.getUiterlijkeEinddatumAfdoening());
        restZoekItem.publicatiedatum = DateTimeConverterUtil.convertToLocalDate(zoekItem.getPublicatiedatum());
        restZoekItem.communicatiekanaal = zoekItem.getCommunicatiekanaal();
        restZoekItem.vertrouwelijkheidaanduiding = zoekItem.getVertrouwelijkheidaanduiding();
        restZoekItem.afgehandeld = zoekItem.isAfgehandeld();
        restZoekItem.groepId = zoekItem.getGroepID();
        restZoekItem.groepNaam = zoekItem.getGroepNaam();
        restZoekItem.behandelaarNaam = zoekItem.getBehandelaarNaam();
        restZoekItem.behandelaarGebruikersnaam = zoekItem.getBehandelaarGebruikersnaam();
        restZoekItem.initiatorIdentificatie = zoekItem.getInitiatorIdentificatie();
        restZoekItem.zaaktypeOmschrijving = zoekItem.getZaaktypeOmschrijving();
        restZoekItem.statustypeOmschrijving = zoekItem.getStatustypeOmschrijving();
        restZoekItem.resultaattypeOmschrijving = zoekItem.getResultaattypeOmschrijving();
        restZoekItem.aantalOpenstaandeTaken = zoekItem.getAantalOpenstaandeTaken();
        restZoekItem.indicatieVerlenging = zoekItem.isIndicatie(ZaakIndicatie.VERLENGD);
        restZoekItem.redenVerlenging = zoekItem.getRedenVerlenging();
        restZoekItem.indicatieOpschorting = zoekItem.isIndicatie(ZaakIndicatie.OPSCHORTING);
        restZoekItem.redenOpschorting = zoekItem.getRedenOpschorting();
        restZoekItem.indicatieDeelzaak = zoekItem.isIndicatie(ZaakIndicatie.DEELZAAK);
        restZoekItem.indicatieHoofdzaak = zoekItem.isIndicatie(ZaakIndicatie.HOOFDZAAK);
        restZoekItem.indicatieHeropend = zoekItem.isIndicatie(ZaakIndicatie.HEROPEND);
        restZoekItem.statusToelichting = zoekItem.getStatusToelichting();
        restZoekItem.indicaties = zoekItem.getZaakIndicaties();
        restZoekItem.rechten = RestRechtenConverter.convert(policyService.readZaakRechten(zoekItem));
        restZoekItem.betrokkenen = new HashMap<>();
        zoekItem.getBetrokkenen().forEach((betrokkenheid, ids) -> {
            restZoekItem.betrokkenen.put(
                    betrokkenheid.replace(ZaakZoekObject.ZAAK_BETROKKENE_PREFIX, ""),
                    ids
            );
        });
        return restZoekItem;
    }
}

File: src/main/kotlin/net/atos/zac/app/zoeken/converter/RestZaakZoekObjectConverter.kt
/*
 * SPDX-FileCopyrightText: 2022 Atos, 2025 Lifely
 * SPDX-License-Identifier: EUPL-1.2+
 */
package net.atos.zac.app.zoeken.converter

import net.atos.zac.app.policy.converter.RestRechtenConverter
import net.atos.zac.app.zoeken.model.RestZaakZoekObject
import net.atos.zac.policy.output.ZaakRechten
import net.atos.zac.util.time.DateTimeConverterUtil.convertToLocalDate
import net.atos.zac.zoeken.model.ZaakIndicatie
import net.atos.zac.zoeken.model.zoekobject.ZaakZoekObject

fun ZaakZoekObject.toRestZaakZoekObject(zaakRechten: ZaakRechten) = RestZaakZoekObject().apply {
    id = this@toRestZaakZoekObject.getObjectId()
    type = this@toRestZaakZoekObject.getType()
    identificatie = this@toRestZaakZoekObject.identificatie
    omschrijving = this@toRestZaakZoekObject.omschrijving
    toelichting = this@toRestZaakZoekObject.toelichting
    archiefNominatie = this@toRestZaakZoekObject.archiefNominatie
    archiefActiedatum = convertToLocalDate(this@toRestZaakZoekObject.archiefActiedatum)
    registratiedatum = convertToLocalDate(this@toRestZaakZoekObject.registratiedatum)
    startdatum = convertToLocalDate(this@toRestZaakZoekObject.startdatum)
    einddatum = convertToLocalDate(this@toRestZaakZoekObject.einddatum)
    einddatumGepland = convertToLocalDate(this@toRestZaakZoekObject.einddatumGepland)
    uiterlijkeEinddatumAfdoening = convertToLocalDate(this@toRestZaakZoekObject.uiterlijkeEinddatumAfdoening)
    publicatiedatum = convertToLocalDate(this@toRestZaakZoekObject.publicatiedatum)
    communicatiekanaal = this@toRestZaakZoekObject.communicatiekanaal
    vertrouwelijkheidaanduiding = this@toRestZaakZoekObject.vertrouwelijkheidaanduiding
    afgehandeld = this@toRestZaakZoekObject.isAfgehandeld
    groepId = this@toRestZaakZoekObject.groepID
    groepNaam = this@toRestZaakZoekObject.groepNaam
    behandelaarNaam = this@toRestZaakZoekObject.behandelaarNaam
    behandelaarGebruikersnaam = this@toRestZaakZoekObject.behandelaarGebruikersnaam
    initiatorIdentificatie = this@toRestZaakZoekObject.initiatorIdentificatie
    zaaktypeOmschrijving = this@toRestZaakZoekObject.zaaktypeOmschrijving
    statustypeOmschrijving = this@toRestZaakZoekObject.statustypeOmschrijving
    resultaattypeOmschrijving = this@toRestZaakZoekObject.resultaattypeOmschrijving
    aantalOpenstaandeTaken = this@toRestZaakZoekObject.aantalOpenstaandeTaken
    indicatieVerlenging = this@toRestZaakZoekObject.isIndicatie(ZaakIndicatie.VERLENGD)
    redenVerlenging = this@toRestZaakZoekObject.redenVerlenging
    indicatieOpschorting = this@toRestZaakZoekObject.isIndicatie(ZaakIndicatie.OPSCHORTING)
    redenOpschorting = this@toRestZaakZoekObject.redenOpschorting
    indicatieDeelzaak = this@toRestZaakZoekObject.isIndicatie(ZaakIndicatie.DEELZAAK)
    indicatieHoofdzaak = this@toRestZaakZoekObject.isIndicatie(ZaakIndicatie.HOOFDZAAK)
    indicatieHeropend = this@toRestZaakZoekObject.isIndicatie(ZaakIndicatie.HEROPEND)
    statusToelichting = this@toRestZaakZoekObject.statusToelichting
    indicaties = this@toRestZaakZoekObject.getZaakIndicaties()
    rechten = RestRechtenConverter.convert(zaakRechten)
    betrokkenen = this@toRestZaakZoekObject.betrokkenen?.mapKeys {
        it.key.replace(ZaakZoekObject.ZAAK_BETROKKENE_PREFIX, "")
    }?.toMutableMap() ?: mutableMapOf()
}
